src/ray/core_worker/task_event_buffer.cc

```diff
@@ -331,6 +426,39 @@ std::unique_ptr<rpc::TaskEventData> TaskEventBufferImpl::CreateDataToSend(
   return data;
 }
 
+void TaskEventBufferImpl::WriteExportData(
+    std::vector<std::shared_ptr<TaskEvent>> &&status_events_to_write_for_export,
+    std::vector<std::shared_ptr<TaskEvent>> &&profile_events_to_send) {
+  absl::flat_hash_map<TaskAttempt, std::shared_ptr<rpc::ExportTaskEventData>>
+      agg_task_events;
+  // Maintain insertion order to agg_task_events so events are written
+  // in the same order as the buffer.
+  std::vector<TaskAttempt> agg_task_event_insertion_order;
+  auto to_rpc_event_fn = [&agg_task_events, &agg_task_event_insertion_order](
+                             std::shared_ptr<TaskEvent> &event) {
+    // Aggregate events by task attempt before converting to proto
+    if (!agg_task_events.count(event->GetTaskAttempt())) {
+      auto inserted = agg_task_events.insert(
+          {event->GetTaskAttempt(), std::make_shared<rpc::ExportTaskEventData>()});
+      RAY_CHECK(inserted.second);
+      agg_task_event_insertion_order.push_back(event->GetTaskAttempt());
+    }
+    auto itr = agg_task_events.find(event->GetTaskAttempt());
+
+    event->ToRpcTaskExportEvents(itr->second);
+  };
+
+  std::for_each(status_events_to_write_for_export.begin(),
+                status_events_to_write_for_export.end(),
+                to_rpc_event_fn);
+  std::for_each(
+      profile_events_to_send.begin(), profile_events_to_send.end(), to_rpc_event_fn);
+
+  for (auto &task_attempt : agg_task_event_insertion_order) {
+    RayExportEvent(agg_task_events[task_attempt]).SendEvent();
```

Comment:
in cpp, it creates an empty class if task_attempt doesn't belong to agg_task_events. So it is recommended to do

```cpp
auto it = agg_task_events.find(task_attempt)
RAY_CHECK(it != agg_task_events.end())
RayExportEvent(it->second).SendEvent()

```